azurerm/internal/services/network/firewall_policy_resource.go

```diff
@@ -0,0 +1,200 @@
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmFirewallPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmFirewallPolicyCreateUpdate,
+		Read:   resourceArmFirewallPolicyRead,
+		Update: resourceArmFirewallPolicyCreateUpdate,
+		Delete: resourceArmFirewallPolicyDelete,
+
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.FirewallPolicyID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.FirewallPolicyName(),
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"location": location.Schema(),
+
+			"base_policy_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validate.FirewallPolicyID,
+			},
+
+			"threat_intelligence_mode": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  string(network.AzureFirewallThreatIntelModeAlert),
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.AzureFirewallThreatIntelModeAlert),
+					string(network.AzureFirewallThreatIntelModeDeny),
+					string(network.AzureFirewallThreatIntelModeOff),
+				}, false),
+			},
+
+			"firewalls": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"rule_collection_groups": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"tags": tags.SchemaEnforceLowerCaseKeys(),
+		},
+	}
+}
+
+func resourceArmFirewallPolicyCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Network.FirewallPoliciesClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if d.IsNewResource() {
+		resp, err := client.Get(ctx, resourceGroup, name, "")
+		if err != nil {
+			if !utils.ResponseWasNotFound(resp.Response) {
+				return fmt.Errorf("checking for existing Firewall Policy %q (Resource Group %q): %+v", name, resourceGroup, err)
+			}
+		}
+
+		if resp.ID != nil && *resp.ID != "" {
+			return tf.ImportAsExistsError("azurerm_firewall_policy", *resp.ID)
+		}
+	}
+
+	param := network.FirewallPolicy{
```

Comment:
Should `param` be `props` here?

Reply:
As this is the root level model instead of the included properties field, hence I'm using the variable name `param` (as what is done in `linux_virtual_machine`)

Reply:
Yep, I realize that, and I understand that `param` has made into the code base some 300 times or so, but for consistency I would really like to see this changed to `props` if you don't mind.

Reply:
Sure, I'll do it.